Migrate to Kotlin.

kenz opened the pull request:
Announced over 50% of professional Android developers now use Kotlin.
So migrated to Kotlin.

kenz committed:
Add Kotlin version.

zhouyiself reviewed (commented):
Hi Kenz,

Thanks so much for your attention to this repo and effort trying to to make it better.

It's definitely right that Kotlin is the tendency and makes sense to provide Kotlin version. While considering there're still a lot of developers who are not familiar with it and feel more comfortable to look into Java implementation, it's *necessary* and valuable to keep the Java version coexist.

This also complies to our convention that has them coexist  (e.g. https://github.com/firebase/quickstart-android/tree/master/mlkit-translate/app/src/main/java/com/google/firebase/samples/apps/mlkit/translate) with the cost of extra maintenance to keep them in sync.

To summarize, let's make the below additional changes in cooperating with your current request:
1. Adds two new sub packages (/java and /kotlin) under the top package com.google.firebase.ml.md for most logic wrote in its corresponding language.
2. Moves few classes (e.g. GraphicOverlay.java) that are used in the layout files to be under the top package in order to not duplicate the layout files.
3. Leaves the MainActivity.java at where it is (top package) and adds entries to the new Kotlin activities.
4. Tweak some auto-generated Kotlin code to meet the guidance here https://developer.android.com/kotlin/style-guide (e.g. several places break the line wrapping style https://developer.android.com/kotlin/style-guide)

I'll do the package movement/restructure if you prefer to not do it in your request.

Appreciated it again for your contribution and let me know your thoughts!

kenz commented:
H, Yi Zhou,

Thank you your feedback for my pull request.

I changed my request.
I tried not duplicate the layout files, however it result was complex code mixed Java and Kotlin.
Because keep independence Java and Kotlin, I copied layout files that  include custom view and modified each file for adapt Java and Kotlin.

Thank you.

rosariopfernandes commented:
Hi @kenz  ,

As @zhouyiself mentioned here:

>     2. Moves few classes (e.g. GraphicOverlay.java) that are used in the layout files to be under the top package in order to not duplicate the layout files.

The classes used in layout files should be under the top package so that we don't have to duplicate layouts, and these layouts can be used in Java and Kotlin both.

kenz commented:
@rosariopfernandes 
Thank you reply.

Yes, I tried it.
However, it became an unreadable code.
The classes used in the Layout file and the it references class (eg CameraSource.java, Size.java and more ...) are either Java or Kotllin.
The capsule was broken(eg com.google.firebase.ml.md.camera and com.google.firebase.ml.md.java.camera and com.google.firebase.ml.md.kotlin.camera).And I had to change some Protected function to Public.

I think so. Many developers need only one of Kotlin or Java. Therefore, it should be easier to sepalated Kotlin and Java.
Duplicating the layout file(only 4 files.) is the easy solution.

rosariopfernandes commented:
@kenz the solution used in [firebase/quickstart-android](https://github.com/firebase/quickstart-android) was to have these classes **written in Java** in a package that could be used by both Java and Kotlin (have a look: https://github.com/firebase/quickstart-android/tree/master/mlkit/app/src/main/java/com/google/firebase/samples/apps/mlkit/common).

I'm not sure if we should keep these classes in Java or convert them to Kotlin. So before you make any changes, I'd suggest we hear @zhouyiself 's thoughts on this.

zhouyiself commented on app/build.gradle:
```diff
@@ -27,14 +30,21 @@ android {
 
     api 'com.google.guava:guava:27.0.1-android'
+    implementation "androidx.core:core-ktx:1.0.1"
+    implementation "androidx.lifecycle:lifecycle-viewmodel-ktx:2.0.0"
```
Nit: move these up to group the androidx ones

zhouyiself commented on app/build.gradle:
```diff
@@ -27,14 +30,21 @@ android {
 }
 
 apply plugin: 'com.google.gms.google-services'
+repositories {
```
Is this really necessary?

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+import com.google.firebase.ml.md.java.Utils;
+
+/**
+ * Entry activity to select the detection mode.
```
s/detection mode/implementation language

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+        super.onCreate(bundle);
+
+        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
+        setContentView(R.layout.activity_entity_choice);
```
activity_entry_choice

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
+        setContentView(R.layout.activity_entity_choice);
+
+        RecyclerView modeRecyclerView = findViewById(R.id.entity_recycler_view);
```
s/modeRecyclerView/entryRecyclerView

zhouyiself replied:
Add also the id name.

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        @Override
+        public void onBindViewHolder(@NonNull EntryItemViewHolder modeItemViewHolder, int position) {
```
entryItemViewHolder

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+        @Override
+        public void onBindViewHolder(@NonNull EntryItemViewHolder modeItemViewHolder, int position) {
+            modeItemViewHolder.bindDetectionMode(entryModes[position]);
+        }
```
bindEntryMode

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/EntryChoiceActivity.java:
```diff
@@ -0,0 +1,121 @@
+                    Activity activity = EntryChoiceActivity.this;
+                    switch (entryMode) {
+                        case ENTRY_JAVA:
+                            activity.startActivity(new Intent(activity, com.google.firebase.ml.md.java.MainActivity.class));
```
Break lines here

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/java/MainActivity.java:
```diff
@@ -132,7 +134,7 @@ void bindDetectionMode(DetectionMode detectionMode) {
                   break;
                 case BARCODE_LIVE:
-                  activity.startActivity(new Intent(activity, LiveBarcodeScanningActivity.class));
+                  activity.startActivity(new Intent(activity, com.google.firebase.ml.md.java.LiveBarcodeScanningActivity.class));
```
Unnecessary package prefix here.

zhouyiself commented on app/src/main/java/com/google/firebase/ml/md/java/Utils.java:
```diff
@@ -53,11 +53,11 @@
   public static final float ASPECT_RATIO_TOLERANCE = 0.01f;
 
-  static final int REQUEST_CODE_PHOTO_LIBRARY = 1;
+  public static final int REQUEST_CODE_PHOTO_LIBRARY = 1;
```
Now that you have forked layout files, let's revert these modifier changes.

zhouyiself commented on build.gradle:
```diff
@@ -1,14 +1,17 @@
-        classpath 'com.google.gms:google-services:4.2.0' // google-services plugin
+        classpath 'com.google.gms:google-services:4.2.0'
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+        // google-services plugin
```
Revert this.

kenz commented:
@rosariopfernandes 
Thank you advise.
I have understand your solution how reduce duplicate layouts with common package.
Sorry, I don't starting work it yet.

zhouyiself reviewed (approved):
Thanks for the contribution!

zhouyiself commented:
Kenz, could we merge these commits into one final commit and described as something like "Adds kotlin version." to keep the commit history more clean without the ones originated from our discussions/reviews?

zhouyiself commented:
Ignore my last comment.

Regarding whether forking layout files, gave it a second thought, it may be more clear to just do the fork for completeness purpose, that is both kotlin and java versions could have the whole implementation package. What do you think?

kenz commented:
@zhouyiself 
Yes, I think, it can more clear to fork layout files.
Each methods have both merits and demerits.
Single layout file take few number of files.
And only need to change single file if when want to change layout file.
Duplicate layout files take to clearly separate to Kotlin and Java.
Each layout files has like behavior, however each files has difference responsibility as sample. so it can keep SRP.

I think that fork layout files is readable, If decide to provide samples for two languages,
